src/lib/projectConfigValidation.ts

```diff
@@ -0,0 +1,439 @@
+      onlyKeyNode = true;
+      message = `Does not match any of the allowed schemas`;
+      break;
+    case AjvErrorKeyword.Pattern:
+      message = `Value does not match pattern: ${error.params.pattern}`;
+      break;
+    case AjvErrorKeyword.Required:
+      onlyKeyNode = true;
+      message = getMissingRequiredFieldError(error.params.missingProperty);
+      break;
+    case AjvErrorKeyword.Type:
+      message = `Incorrect type: ${error.message}`;
+      break;
+    default:
+      break;
+  }
+  return { message, nodePath, onlyKeyNode };
+}
+
+function diagnosticDeduplicationKey(
+  nodePath: (string | number)[],
+  keyword: string,
+  params: any
+): string {
+  return `${keyword}::${nodePath.join('/')}::${JSON.stringify(params)}`;
+}
+
+function getKeyNode(node: jsonc.Node): jsonc.Node | undefined {
+  return node.parent?.type === 'property' ? node.parent.children?.[0] : node;
+}
+
+async function getCachedSchemasForAllTypes(
+  projectRoot: Uri,
+  platformVersion: string,
+  accountId: number
+): Promise<SchemaCacheEntry> {
+  // If schema cache has a schema for this version, return it
+  // If there is an error, we're going to return that too
+  if (schemaCache[platformVersion]) {
+    return schemaCache[platformVersion];
+  }
+  let response: Record<string, AnySchema>;
+  try {
+    response = await getIntermediateRepresentationSchema({
+      projectSourceDir: projectRoot.fsPath,
+      platformVersion,
+      accountId,
+    });
+    let byType: Record<string, SchemaAndValidator> = {};
+    for (const [key, schema] of Object.entries(response)) {
+      byType[key] = { schema };
+    }
+    const result = {
+      byType,
+    };
+    schemaCache[platformVersion] = result;
+    return result;
+  } catch (error) {
+    const message = `${getFailedToFetchSchemasError()}. Reload your window to try again`;
+    console.error(message, error);
+    const result = {
+      error: message,
+    };
+    schemaCache[platformVersion] = result;
+    return result;
+  }
+}
+
+async function getCachedValidatorForExternalType(
+  projectRoot: Uri,
+  platformVersion: string,
+  accountId: number,
+  externalType: string
+): Promise<ValidateFunction | { error: string }> {
+  const cachedSchemas = await getCachedSchemasForAllTypes(
+    projectRoot,
+    platformVersion,
+    accountId
+  );
+  if ('error' in cachedSchemas) {
+    return cachedSchemas;
+  }
+  const internalType = mapToInternalType(externalType);
+  if (!(internalType in cachedSchemas.byType)) {
+    return { error: getUnsupportedTypeError(externalType) };
+  }
+  const schemaAndValidator = cachedSchemas.byType[internalType];
+  let validator = schemaAndValidator.validator;
+  if (!validator) {
+    validator = ajv.compile(schemaAndValidator.schema);
+    schemaAndValidator.validator = validator;
+  }
+  return validator;
+}
+
+// VSCode doesn't provide a dirname function, that's in the github:/microsoft/vscode-uri package
+function vscodeDirname(uri: Uri): Uri {
```

Comment:
Nit: We could rename this to `dirname` and maybe move it to the `fileHelpers .ts`